hudi-common/src/main/java/org/apache/hudi/common/engine/HoodieReaderContext.java

```diff
@@ -72,15 +73,16 @@ public void setSchemaHandler(HoodieFileGroupReaderSchemaHandler<T> schemaHandler
     this.schemaHandler = schemaHandler;
   }
 
-  public String getTablePath() {
-    if (tablePath == null) {
-      throw new IllegalStateException("Table path not set in reader context.");
-    }
-    return tablePath;
+  public StoragePath getTablePath() {
+    return metaClient.getBasePath();
+  }
+
+  public HoodieTableMetaClient getMetaClient() {
+    return metaClient;
   }
 
-  public void setTablePath(String tablePath) {
-    this.tablePath = tablePath;
+  public void setMetaClient(HoodieTableMetaClient metaClient) {
```

Comment:
`HoodieReaderContext` is meant to handle only engine-specific implementation on reading data files, getting field values from a record, transforming a record, etc., but now it's getting more and more complex and hard to maintain by mixing Hudi-specific logic which should be completely avoided.  Keep this in mind.  If you need to have common logic for Hudi-specific logic, let's add a new class that can be used across engines without any changes, just like `HoodieFileGroupReader`, and keep this class lean.

Reply:
Could you file a ticket on this as a follow-up?

Reply:
Rethinking this, should we keep meta client out of the reader context?